File: solutions/Leetcode_120/leetcode_120.py
from typing import List


class Solution1:
    def minimum_total(self, triangle: List[List[int]]) -> int:
        level = triangle[0]
        for i in range(1, len(triangle)):
            temp = triangle[i]
            temp[0] += level[0]
            temp[-1] += level[-1]
            for j in range(1, i):
                temp[j] += min(level[j-1], level[j])
            level = temp
        return min(level)


class Solution2:
    def minimum_total(self, triangle):
        level = triangle[-1]
        for i in range(len(triangle)-2, -1, -1):
            for j in range(i+1):
                level[j] = triangle[i][j] + min(level[j], level[j+1])
        return level[0]


if __name__ == '__main__':
    triangle = [[2], [3, 4], [6, 5, 7], [4, 1, 8, 3]]
    ans = Solution2().minimum_total(triangle)
    print(ans)

File: solutions/Leetcode_120/leetcode_120.cc
#include <vector>
#include <bits/stdc++.h>

using namespace std;

class Solution1
{
public:
    int minimumTotal(vector<vector<int>> &triangle)
    {
        vector<int> level = triangle[0];
        for (int i = 1; i != triangle.size(); ++i)
        {
            vector<int> temp = triangle[i];
            temp[0] += level[0];
            temp[temp.size() - 1] += level[level.size() - 1];
            for (int j = 1; j != temp.size() - 1; ++j)
                temp[j] += min(level[j - 1], level[j]);
            level = temp;
        }
        return *min_element(level.begin(), level.end());
    }
};

class Solution2
{
public:
    int minimumTotal(vector<vector<int>> &triangle)
    {
        vector<int> level = triangle.back();
        for (int i = triangle.size() - 2; i != -1; --i)
            for (int j = 0; j != triangle[i].size(); ++j)
                level[j] = triangle[i][j] + min(level[j], level[j + 1]);
        return level[0];
    }
};